runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java

```diff
@@ -0,0 +1,376 @@
+
+    public LayoutManagementView(@NonNull CorfuRuntime runtime) {
+        super(runtime);
+    }
+
+    private volatile long prepareRank = 1L;
+
+    /**
+     * On restart, if MANAGEMENT_LAYOUT exists in the local datastore.
+     * the Management Server attempts to recover the cluster from that layout.
+     *
+     * @param recoveryLayout Layout to use to recover
+     * @return True if the cluster was recovered, False otherwise
+     */
+    public void recoverCluster(Layout recoveryLayout)
+            throws QuorumUnreachableException, OutrankedException, InterruptedException,
+            ExecutionException, LayoutModificationException, CloneNotSupportedException {
+
+        Layout newLayout = (Layout) recoveryLayout.clone();
+        newLayout.setEpoch(recoveryLayout.getEpoch() + 1);
+        newLayout.setRuntime(runtime);
+        runLayoutReconfiguration(recoveryLayout, newLayout, true);
+    }
+
+    /**
+     * Takes in the existing layout and a set of failed nodes.
+     * It first generates a new layout by removing the failed nodes from the existing layout.
+     * It then seals the epoch to prevent any client from accessing the stale layout.
+     * Finally we run paxos to update all servers with the new layout.
+     *
+     * @param currentLayout The current layout
+     * @param failedServers Set of failed server addresses
+     * @param healedServers Set of healed server addresses
+     */
+    public void handleFailure(IFailureHandlerPolicy failureHandlerPolicy,
+                              Layout currentLayout,
+                              Set<String> failedServers,
+                              Set<String> healedServers)
+            throws QuorumUnreachableException, OutrankedException, InterruptedException,
+            ExecutionException, LayoutModificationException, CloneNotSupportedException {
+
+        // Generates a new layout by removing the failed nodes from the existing layout
+        Layout newLayout = failureHandlerPolicy
+                .generateLayout(currentLayout,
+                        runtime,
+                        failedServers,
+                        healedServers);
+        runLayoutReconfiguration(currentLayout, newLayout, false);
+    }
+
+
+    /**
+     * Bootstraps the new node with the current layout.
+     * This action is invoked as a part of the add node workflow which requires the new node
+     * to be added to the cluster to be bootstrapped with the existing layout of this cluster.
+     * This bootstraps the Layout and the Management server with the existing layout and also
+     * initiates failure handling capabilities on the management server.
+     *
+     * @param endpoint New node endpoint.
+     * @throws InterruptedException       if bootstrapping interrupted.
+     * @throws ExecutionException         if bootstrapping fails.
+     * @throws CloneNotSupportedException if layout clone fails.
+     */
+    public void bootstrapNewNode(String endpoint)
```

Comment:
This should also throw  AlreadyBootstrappedException

Reply:
Done